src/ast/passes/semantic_analyser.cpp

```diff
@@ -1334,6 +1334,40 @@ void SemanticAnalyser::visit(Call &call)
     }
     call.type = CreateUInt32();
   }
+  else if (call.func == "printb")
+  {
+    // only support one arguement
+    if (!check_nargs(call, 1))
+      return;
+
+    Expression *arg = call.vargs->at(0);
+    if (arg->type.type == Type::pointer)
```

Comment:
I feel pointers could be better. Because from the scripting point of view, it will be a little simpler. For example printb(skb->sk) instead of printb(*(skb->sk)).  What do you think ?

Reply:
If you go with structs, it will allow to use `printb` in more use-cases, e.g. `@ = *($skb->sk); ... printb(@)`. Also, it'll be easier implementation-wise b/c now you're always doing proberead to access the memory but that's not always the correct way (e.g. BTF-based probes don't need it).

Reply:
> If you go with structs, it will allow to use printb in more use-cases, e.g. @ = *($skb->sk); ... printb(@).
In fact, it is now supported. `@=$skb->sk; ... printb(@);.

> Also, it'll be easier implementation-wise b/c now you're always doing proberead to access the memory but that's not always the correct way (e.g. BTF-based probes don't need it).

I suddenly remembered why I can't dereference directly, because the general kernel structure is relatively large, and if I dereference it directly, it will occupy the eBPF stack space.

Reply:
> In fact, it is now supported. `@=$skb->sk; ... printb(@);.

True, but that requires kernel to be able to store pointers into maps which hasn't been around for long.
 
> I suddenly remembered why I can't dereference directly, because the general kernel structure is relatively large, and if I dereference it directly, it will occupy the eBPF stack space.

IIUC, proberead will also read the data onto BPF stack so there's no difference on that front.

Since `printb` is printing the structure, it still feels natural to me to pass directly the structure and not a pointer to the call.

Reply:
But I also see your arguments. Would be nice to get opinions from others here.

Reply:
> > In fact, it is now supported. `@=$skb->sk; ... printb(@);.
> 
> True, but that requires kernel to be able to store pointers into maps which hasn't been around for long.
> 
> > I suddenly remembered why I can't dereference directly, because the general kernel structure is relatively large, and if I dereference it directly, it will occupy the eBPF stack space.
> 
> IIUC, proberead will also read the data onto BPF stack so there's no difference on that front.
>

We can directly use proberead to read the data into the map. The prototype of proberead is `long bpf_probe_read(void *dst, u32 size, const void *unsafe_ptr)`, we can set dst as the value pointer of map.

> Since `printb` is printing the structure, it still feels natural to me to pass directly the structure and not a pointer to the call.

Reply:
> We can directly use proberead to read the data into the map. The prototype of proberead is `long bpf_probe_read(void *dst, u32 size, const void *unsafe_ptr)`, we can set dst as the value pointer of map.

Fair point, I didn't realize that. Could we do the same for BTF-based probes by directly loading into the map?

Also, I'm wondering if LLVM would be able to optimize away the stack usage even if we had it in the codegen. If not, perhaps it'd make sense to support both variants (pointer and non-pointer)?

Reply:
> > We can directly use proberead to read the data into the map. The prototype of proberead is `long bpf_probe_read(void *dst, u32 size, const void *unsafe_ptr)`, we can set dst as the value pointer of map.
> 
> Fair point, I didn't realize that. Could we do the same for BTF-based probes by directly loading into the map?
>

 I understand it should be possible, we can even use percpu map instead of eBPF stack.

> Also, I'm wondering if LLVM would be able to optimize away the stack usage even if we had it in the codegen. If not, perhaps it'd make sense to support both variants (pointer and non-pointer)?

Sorry, I don't quite understand the meaning here.:)

Reply:
The question was that if we have something like this in codegen:
```
get pointer to skb
proberead skb to BPF stack
store the data (from BPF stack) into printb map
```
if LLVM optimizations are able to turn it into
```
get pointer to skb
proberead skb to printb map
```

If not, then, we'll have to do the optimization on the level of the `printb` call and then, passing a pointer seems a better option implementation-wise. On the other hand, user-wise, it seems more appropriate to pass a structure so I'm thinking if we could allow both.

Reply:
I understand what you mean, llvm should not have this optimization, because the two of them are completely independent memory spaces.